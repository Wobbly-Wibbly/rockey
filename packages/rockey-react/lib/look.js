import rule from 'rockey/rule';
import when from 'rockey/when';
import insert from 'rockey/insert';
import condition from 'rockey/condition';

import htmlTags from './htmlTags';

import RockeyHoc from './';

const look = BaseComponent =>
  (...args) => {
    const css = rule(...args);

    return css.transform((tree, create) => {
      const components = Object.keys(tree.components);
      const size = components.length;

      if (!size) {
        throw new Error(
          'Rockey look should containt at least one root component'
        );
      }

      const parentDisplayName = components[0];
      const parentCss = create({
        components: {
          [parentDisplayName]: tree.components[parentDisplayName],
        },
      });

      const children = {};

      for (let i = 1; i < size; i++) {
        const displayName = components[i];

        const comopnentCss = create({
          components: {
            [displayName]: tree.components[displayName],
          },
        });
        comopnentCss.addParent(parentCss);

        children[displayName] = RockeyHoc(BaseComponent, {
          displayName,
          parentCss: comopnentCss,
        });
      }

      return {
        ...children,
        [parentDisplayName]: RockeyHoc(BaseComponent, {
          displayName: parentDisplayName,
          parentCss: parentCss,
        }),
      };
    });
  };

look.when = when;
look.condition = condition;
look.insert = insert;

for (const tag of htmlTags) {
  // ---- tag hoc lazy creation
  Object.defineProperty(look, tag, {
    get: () => (...args) => look(tag)(...args),
  });
}

export default look;
