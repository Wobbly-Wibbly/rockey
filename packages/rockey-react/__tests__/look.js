import React from 'react';
import isFunction from 'lodash/isFunction';
import renderer from 'react-test-renderer';

import look from '../lib/look';

jest.mock('rockey/utils/hash', () => {
  return () => '{{ hash }}';
});

describe('look', () => {
  it('when', () => {
    expect(isFunction(look.when)).toBeTruthy();
  });

  it('condition', () => {
    expect(isFunction(look.condition)).toBeTruthy();
  });

  it('insert', () => {
    expect(isFunction(look.insert)).toBeTruthy();
  });
});

test('Look without components', () => {
  expect(() => look.button``).toThrow(
    'Rockey look should containt at least one root component'
  );
});

test('Look shortcuts', () => {
  const { Button, PrimaryButton } = look.button`
    Button {
      color: red;
      font-size: bold;
    }

    PrimaryButton {
      color: blue;
    }
  `;

  const ButtonTree = renderer.create(<Button>Button</Button>).toJSON();
  expect(ButtonTree).toMatchSnapshot();

  const PrimaryButtonTree = renderer
    .create(<PrimaryButton>PrimaryButton</PrimaryButton>)
    .toJSON();
  expect(PrimaryButtonTree).toMatchSnapshot();
});

test('Look', () => {
  const { Layer, PrimaryLayer } = look(({ children, className }) => (
    <div className={className}>{children}</div>
  ))`
    Layer {
      padding: 10px;
      border: 1px solid #000;
      color: green;
    }

    PrimaryLayer {
      color: red;
    }
  `;

  expect(isFunction(Layer)).toBeTruthy();
  expect(isFunction(PrimaryLayer)).toBeTruthy();

  const LayerTree = renderer.create(<Layer>Layer</Layer>).toJSON();
  expect(LayerTree).toMatchSnapshot();

  const PrimaryLayerTree = renderer
    .create(<PrimaryLayer>PrimaryLayer</PrimaryLayer>)
    .toJSON();
  expect(PrimaryLayerTree).toMatchSnapshot();
});
