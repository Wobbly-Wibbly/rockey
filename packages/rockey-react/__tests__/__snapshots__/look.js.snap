// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Look 1`] = `
<div
  className="Layer-{{ hash }}"
>
  Layer
</div>
`;

exports[`Look 2`] = `
<div
  className="PrimaryLayer-{{ hash }} Layer-{{ hash }}"
>
  PrimaryLayer
</div>
`;

exports[`Look 3`] = `
<div
  className="PrimaryLayer-{{ hash }} Layer-{{ hash }} Mixin-anonWhen-{{ hash }}-1"
>
  PrimaryLayer
</div>
`;

exports[`Look shortcuts 1`] = `
<button
  className="Button-{{ hash }}"
>
  Button
</button>
`;

exports[`Look shortcuts 2`] = `
<button
  className="PrimaryButton-{{ hash }} Button-{{ hash }}"
>
  PrimaryButton
</button>
`;
